pallets/rewards/src/mechanism/gap.rs

```diff
@@ -0,0 +1,377 @@
+				group.rpt.ensure_add_assign(rate)?;
+				reward = amount;
+			}
+
+			group
+				.total_stake
+				.ensure_add_assign(group.pending_total_stake)?;
+			group.pending_total_stake = T::Balance::zero();
+
+			RptHistory::<T>::insert(group.distribution_id, group.rpt);
+
+			group.distribution_id = LastDistributionId::<T>::try_mutate(
+				|distribution_id| -> Result<T::DistributionId, DispatchError> {
+					Ok(*distribution_id.ensure_add_assign(One::one())?)
+				},
+			)?;
+
+			Ok(reward)
+		}
+
+		fn deposit_stake(
+			account: &mut Self::Account,
+			currency: &mut Self::Currency,
+			group: &mut Self::Group,
+			amount: Self::Balance,
+		) -> Result<(), DispatchError> {
+			account.update(group, currency)?;
+
+			account.pending_stake.ensure_add_assign(amount)?;
+			group.pending_total_stake.ensure_add_assign(amount)?;
+
+			currency.total_stake.ensure_add_assign(amount)?;
+
+			Ok(())
+		}
+
+		fn withdraw_stake(
+			account: &mut Self::Account,
+			currency: &mut Self::Currency,
+			group: &mut Self::Group,
+			amount: Self::Balance,
+		) -> Result<(), DispatchError> {
+			account.update(group, currency)?;
+
+			let pending_amount = amount.min(account.pending_stake);
+
+			account.pending_stake.ensure_sub_assign(pending_amount)?;
+			group
+				.pending_total_stake
+				.ensure_sub_assign(pending_amount)?;
+
+			let computed_amount = amount.ensure_sub(pending_amount)?;
+
+			account.stake.ensure_sub_assign(computed_amount)?;
+			account
+				.reward_tally
+				.ensure_sub_assign(group.rpt.ensure_mul_int(computed_amount)?.ensure_into()?)?;
+			group.total_stake.ensure_sub_assign(computed_amount)?;
+
+			currency.total_stake.ensure_sub_assign(amount)?;
+
+			Ok(())
+		}
+
+		fn compute_reward(
+			account: &Self::Account,
+			currency: &Self::Currency,
+			group: &Self::Group,
+		) -> Result<Self::Balance, DispatchError> {
+			let stake = account.stake_updated(group, currency)?;
+			let reward_tally = account.reward_tally_updated(group, currency)?;
+
+			T::IBalance::ensure_from(group.rpt.ensure_mul_int(stake)?)?
+				.ensure_sub(reward_tally)?
+				.ensure_into()
+				.map_err(|e| e.into())
+		}
+
+		fn claim_reward(
+			account: &mut Self::Account,
+			currency: &Self::Currency,
+			group: &Self::Group,
+		) -> Result<Self::Balance, DispatchError> {
+			account.update(group, currency)?;
+
+			let reward = Self::compute_reward(&account, currency, group)?;
+
+			account
+				.reward_tally
+				.ensure_add_assign(reward.ensure_into()?)?;
+
+			Ok(reward)
+		}
+
+		fn move_currency(
+			currency: &mut Self::Currency,
+			prev_group: &mut Self::Group,
+			next_group: &mut Self::Group,
+		) -> Result<(), MoveCurrencyError> {
+			if prev_group.pending_total_stake > T::Balance::zero() {
+				Err(DispatchError::from(
+					Error::<T>::TryMovementAfterPendingState,
+				))?;
+			}
+
+			let rpt_change = next_group.rpt.ensure_sub(prev_group.rpt)?;
+
+			currency
+				.rpt_changes
+				.try_push(rpt_change)
```

Comment:
It is possible that the `rpt_change` is negative. Is that Arithmetically handled fine by the consumer of `currency.rpt_changes`?

Reply:
Good question! Yes, it is possible if you move to a group with less rpt. It implies that the `Fixed` type used for `Rate` should be a signed type to handle this. The guard `<T::Rate as FixedPointNumber>::Inner: Signed` in the `RewardMechanism` impl protects this from a wrong configuration, with a `FixedU64` for example.